doozerlib/cli/release_gen_payload.py

```diff
@@ -58,762 +145,869 @@ def modify_and_replace_api_object(api_obj: oc.APIObject, modifier_func: Callable
+                stream=ml, default_flow_style=False)
+        exectools.cmd_assert(f"manifest-tool push from-spec {str(component_manifest_path)}", retries=3)
+
+        # we are pushing a new manifest list, so return its sha256 based pullspec
+        return exchange_pullspec_tag_for_shasum(output_pullspec, find_manifest_list_sha(output_pullspec))
+
+    def create_multi_release_image(
+            self, imagestream_name: str, multi_release_is: Dict, multi_release_dest: str,
+            multi_release_name: str, multi_specs: Dict[bool, Dict[str, Dict[str, PayloadEntry]]],
+            private_mode: bool) -> Dict[str, str]:
+        """
+        Create and publish a "multi" release image for each arch. These all have the same content,
+        just the release image itself is arch-specific (based on arch CVO). Then stitch them
+        together into a manifest list.
+        Returns: a pullspec for the manifest list of release images.
+        """
+        # Write the imagestream to a file ("oc adm release new" can read from a file instead of
+        # openshift cluster API)
+        multi_release_is_path: Path = self.output_path.joinpath(f"{imagestream_name}-release-imagestream.yaml")
+        with multi_release_is_path.open(mode="w+") as mf:
             yaml.safe_dump(multi_release_is, mf)
 
-        multi_release_dest = f'quay.io/{organization}/{release_repository}:{multi_release_manifest_list_tag}'
         arch_release_dests: Dict[str, str] = dict()  # This will map arch names to a release payload pullspec we create for that arch (i.e. based on the arch's CVO image)
-        for arch, payload_entry in multi_specs[private_mode]['cluster-version-operator'].items():
-            cvo_pullspec = payload_entry.dest_pullspec
-            arch_release_dest = f'{multi_release_dest}-{arch}'
-            arch_release_dests[arch] = arch_release_dest
-
-            new_release_cmd = [
-                'oc',
-                'adm',
-                'release',
-                'new',
-                f'--name={multi_release_name}',
-                '--reference-mode=source',
-                '--keep-manifest-list',
-                f'--from-image-stream-file={str(multi_release_is_path)}',
-                f'--to-image-base={cvo_pullspec}',
-                f'--to-image={arch_release_dest}'
-            ]
-
-            metadata: Dict[str, str] = dict()
-            metadata['release.openshift.io/architecture'] = 'multi'
-            new_release_cmd.append("--metadata")
-            new_release_cmd.append(json.dumps(metadata))
-
-            # Create the arch specific release payload containing tags pointing to manifest list component images.
-            exectools.cmd_assert(new_release_cmd)
-
-        # Create manifest list spec containing references to all the arch specific release payloads we've created
-        manifests = []
+        for arch, payload_entry in multi_specs[private_mode]["cluster-version-operator"].items():
```

Comment:
I noticed in a test run limited to two images that if I left out `cluster-version-operator` this failed (of course). But it occurred to me to wonder whether we get a multi payload with only the listed group items. I think so, since there's no way to retrieve the set of tags that were synced in previous runs; well, payload creation probably fails due to missing references but... maybe not if components are included by label instead of reference? There should probably be safeguards against syncing a partial hetero payload.

Reply:
(after discussion)
If we had to, we could look up the existing single-arch imagestreams, stitch them together, and then overlay them with the targeted content updates. However this seems like a lot of work we probably won't need. The simpler solution is to disallow this in the first place. When the image set is limited (either by inclusion or exclusion) and a multi release is requested, raise an error during the doozer parameter validation. Params should either enable a full multi release or not request a multi release.

This may be a follow-on improvement.